src/Webapi/Canvas/Webapi__Canvas__CanvasImageSource.res

```diff
@@ -0,0 +1,51 @@
+type t
+
+module Impl = (
+  T: {
+    type t
+  },
+) => {
+  type t_canvasImageSource = T.t
+
+  %%private(
+    @send
+    external drawImageJS: (
+      Webapi__Canvas__Canvas2d.t,
+      ~image: t_canvasImageSource,
+      ~dx: float,
+      ~dy: float,
+    ) => unit = "drawImage"
+
+    @send
+    external drawImageScaleJS: (
+      Webapi__Canvas__Canvas2d.t,
+      ~image: t_canvasImageSource,
+      ~dx: float,
+      ~dy: float,
+      ~dWidth: float,
+      ~dHeight: float,
+    ) => unit = "drawImage"
+
+    @send
+    external drawImageFullJS: (
+      Webapi__Canvas__Canvas2d.t,
+      t_canvasImageSource,
+      float,
+      float,
+      float,
+      float,
+      float,
+      float,
+      float,
+      float,
+    ) => unit = "drawImage"
+  )
+
+  let drawImage = (image, ~ctx, ~dx, ~dy) => ctx->drawImageJS(~image, ~dx, ~dy)
+
+  let drawImageScale = (image, ~ctx, ~dx, ~dy, ~dWidth, ~dHeight) =>
+    ctx->drawImageScaleJS(~image, ~dx, ~dy, ~dWidth, ~dHeight)
+
+  let drawImageFull = (image, ~ctx, ~sx, ~sy, ~sWidth, ~sHeight, ~dx, ~dy, ~dWidth, ~dHeight) =>
+    ctx->drawImageFullJS(image, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight)
```

Comment:
Why have you structured it in this way? I disagree with making the `image` the focus of the bindings. It's a Canvas2d method, it will be expected to be in the `Canvas2d` module.

I don't see how `image->HtmlImageElement.drawImage(~ctx, ~dx=0.0, ~dy=0.0)` is any better than `ctx->Canvas2d.drawImage(image, ~dx=0.0, ~dy=0.0)`. It also removes the runtime code generated by this technique (note all the `Curry` calls in your test JS).

I'm happy to discuss if there is a good reason to offer `drawImage` on the `HtmlImageElement` module.

Reply:
I agree 100% and is what I originally wanted.  Only issues is that I an not sure how to get the typing to work out as image could also be a bunch of other types.

Other things that image could be:
- HtmlImageElement
- SvgImageElement
- HtmlVideoElement
- ImageBitmap
- OffscreenCanvas

Reply:
This module should only have identity functions converting the various things into the opaque `t_imageSource`. As Andy said the draw function bindings should be in the `Webapi__Canvas__Canvas2d.res`

another alternative would be to have convenience bindings for each of these types but it would end up being a lot of them since we have 3 types of arguments variants times the number of formats.

So something like this:
`ctx->Canvas2d.drawImage(image->CanvasImageSource.imageElementToImageSource, ~dx=0.0, ~dy=0.0)'
'ctx->Canvas2d.drawImage(svg->CanvasImageSource.svgElementToImageSource, ~dx=0.0, ~dy=0.0)`

Reply:
Ah I didn't read the spec closely enough to see there could be so many inputs. That makes sense. And I see now why you only made one - that's the only one that exists in the webapi bindings today.

To keep the names short, things like `CanvasImageSource.ofImage` and `CanvasImageSource.ofSvg` would be fine. But perhaps we go the other way, and avoid needing a `CanvasImageSource` module completely by adding the methods to each element module?
```
ctx->Canvas2d.drawImage(image->HtmlImageElement.asCanvasSource, ~dx=0.0, ~dy=0.0)
ctx->Canvas2d.drawImage(svg->SvgImageElement.asCanvasSource, ~dx=0.0, ~dy=0.0)
```

Reply:
Maybe we should then split out the BlobPart things into it's own module similar to this. Currently they are in the same format as the one I suggested.
https://github.com/tinymce/rescript-webapi/blob/main/src/Webapi/Webapi__Blob.res#L39

I think it is better to have the convert being in the module that needs to consume it instead of having it spread out since for the Blob case some of those inputs are not in the webapi like Uint8Array is part of the rescript core bindings. This would keep it more consitent.

So this for the canvas bindings
`ctx->Canvas2d.drawImage(image->CanvasImageSource.ofImage, ~dx=0.0, ~dy=0.0)`

And we change the blob part to be it's own similar module like this
`Blob.make([bytes->BlobPart.ofUint8Array])`

Then both of these would be similar to `Element.ofNode`.

Reply:
Ah, I see what you mean. Alright, let's go with `CanvasImageSource.ofImage`.